### PR TITLE
Expression editor test capability timeout

### PR DESCRIPTION
**Pull request by ilasalexandra**
- Added new method for testing capability.
- Refactored timeout logic. 

Signed-off-by: Alexandra Ilas alexandra.ilas@microfocus.com

### CODE REVIEW

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionNotCachedEngine.java [hunk 1]
```diff
@@ -52,4 +52,13 @@ public PythonEvaluationResult eval(String prepareEnvironmentScript, String scrip
...
     }
+
+    @Override
+    public PythonEvaluationResult test(String prepareEnvironmentScript, String script, Map<String, Serializable> vars, long timeout) {
+        PythonExecutor pythonExecutor = new PythonExecutor(Collections.<String>emptySet());
+        try {
+            return pythonExecutor.eval(prepareEnvironmentScript, script, vars);
+        } finally {
+            pythonExecutor.close();
+        }    }
```
Comment (alexbajzat) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionNotCachedEngine.java:63]: am i seeing double ? `}      }`

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 2]
```diff
@@ -126,6 +127,28 @@ public PythonEvaluationResult eval(String expression, String prepareEnvironmentS
...
     }
 
+    public PythonEvaluationResult test(String expression, String prepareEnvironmentScript,
+                                       Map<String, Serializable> context, long timeout) {
+        TempEvalEnvironment tempEvalEnvironment = null;
+        try {
+            String pythonPath = checkPythonPath();
+            tempEvalEnvironment = generateTempResourcesForEval();
+            String payload = generatePayloadForEval(expression, prepareEnvironmentScript, context);
+            addFilePermissions(tempEvalEnvironment.parentFolder);
```
Comment (alexbajzat) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:137]: This code is probably duplicated with the `eval`'s method one. Can you please extract it in a private method ?

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 3]
```diff
@@ -190,7 +213,30 @@ private PythonEvaluationResult runPythonEvalProcess(String pythonPath, String pa
...
+                throw new ExternalPythonEvalException(exception);
+            }
+            context.put("accessed_resources_set", (Serializable) scriptResults.getAccessedResources());
+            //noinspection unchecked
+            return new PythonEvaluationResult(processReturnResult(scriptResults), context);
+        } catch (IOException | InterruptedException | ExecutionException | TimeoutException e) {
+            logger.error("Failed to run script. ", e.getCause() != null ? e.getCause() : e);
+            throw new RuntimeException("Failed to run script.");
+        }
+    }
```
Comment (alexbajzat) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:231]: All this code si the same in the test method, except for the timeout value and the exception handling. Can you do one method for both and just leverage between eval/test ? Maybe just send the timeout value and you can throw the same timeout exception message for eval as well.

**Review by alexbajzat — CHANGES_REQUESTED** [review]

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionNotCachedEngine.java (hunk 1 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionNotCachedEngine.java]: done

**Review by alexbajzat — APPROVED** [review]

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java (hunk 2 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java]: done

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java (hunk 3 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java]: done

**Review by lucian-cm — CHANGES_REQUESTED** [review]

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionCachedEngine.java [hunk 4]
```diff
@@ -58,6 +58,16 @@ public PythonEvaluationResult eval(String prepareEnvironmentScript, String scrip
         }
     }
 
+    @Override
+    public PythonEvaluationResult test(String prepareEnvironmentScript, String script, Map<String, Serializable> vars, long timeout) {
+        PythonExecutor executor = allocateExecutor(Sets.<String>newHashSet());
```
[the remark this answers is not in the document]
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionCachedEngine.java]: done

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 5]
```diff
@@ -83,6 +86,25 @@
...
+        engineExecutorService = Executors.newFixedThreadPool(10, threadFactory);
+
+    }
+
+    static {
+        final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat("python-test-%d")
+                .setDaemon(true)
+                .build();
+        testExecutorService = Executors.newFixedThreadPool(3, threadFactory);
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:103]: 3 needs to be a exposed as a system property with default

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 6]
```diff
@@ -83,6 +86,25 @@
         objectMapper = new ObjectMapper(factory);
     }
 
+    static {
+        final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat("python-engine-%d")
+                .setDaemon(true)
+                .build();
+        engineExecutorService = Executors.newFixedThreadPool(10, threadFactory);
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:94]: 10 needs to be a exposed as a system property with default

File: runtime-management/runtime-management-impl/src/test/java/io/cloudslang/runtime/impl/python/PythonExecutorTest.java [hunk 7]
```diff
@@ -330,7 +330,7 @@ public PythonRuntimeService pythonRuntimeService() {
 
         @Bean(name = "externalPythonRuntimeService")
         public PythonRuntimeService externalPythonRuntimeService() {
-            return new ExternalPythonRuntimeServiceImpl(new Semaphore(100));
+            return new ExternalPythonRuntimeServiceImpl(new Semaphore(100),new Semaphore(50));
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/test/java/io/cloudslang/runtime/impl/python/PythonExecutorTest.java:333]: missing whitespace after comma

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java [hunk 8]
```diff
@@ -90,4 +93,30 @@ public PythonEvaluationResult eval(String prepareEnvironmentScript, String scrip
...
+                    testingControlSemaphore.release();
+                }
+            } else {
+                logger.warn("Maximum number of python processes has been reached. Waiting for a python process to finish. " +
+                        "You can configure the number of concurrent python executions by setting " +
+                        "'python.testing.concurrent.execution.permits' system property.");
+                testingControlSemaphore.acquire();
+                try {
+                    logger.info("Acquired a permit for a new python process. Continuing with execution...");
+                    return externalPythonExecutionEngine.test(prepareEnvironmentScript, script, vars,timeout);
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java:113]: check for bad formatting, near timeout

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 9]
```diff
@@ -218,34 +251,31 @@ private Serializable processReturnResult(EvaluationResults results) throws JsonP
...
+        try {
+            return CompletableFuture.supplyAsync(supplier, executorService).get(timeoutMillis, MILLISECONDS);
+        } catch (TimeoutException e) {
+            try {
+                Process process = supplier.getProcess();
+                if (process != null) {
+                    process.destroy();
+                }
+            } catch (Exception t) {
+                logger.error("Failed to run script. ", t.getCause() != null ? t.getCause() : t);
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:273]: should swallow this exception as we need to throw RuntimeException with TimeoutException as cause.

**Review by lucian-cm — CHANGES_REQUESTED** [review]

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionEngineConfiguration.java [hunk 10]
```diff
@@ -42,7 +42,8 @@ public PythonRuntimeService pythonRuntimeService() {
     @Bean(name = "externalPythonRuntimeService")
     public PythonRuntimeService externalPythonRuntimeService() {
         Integer pythonProcessPermits = Integer.getInteger("python.concurrent.execution.permits", 30);
-        return new ExternalPythonRuntimeServiceImpl(new Semaphore(pythonProcessPermits));
+        Integer pythonTestingProcessPermits = Integer.getInteger("python.testing.concurrent.execution.permits", 10);
+        return new ExternalPythonRuntimeServiceImpl(new Semaphore(pythonProcessPermits),new Semaphore(pythonTestingProcessPermits));
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/PythonExecutionEngineConfiguration.java:46]: whitespace issue after comma

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonEvaluationSupplier.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            printWriter.flush();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+            String line;
+            StringBuilder returnResult = new StringBuilder();
+            while ((line = reader.readLine()) != null) {
+                returnResult.append(line);
+            }
+            return returnResult.toString();
+        } catch (IOException e) {
+            throw new RuntimeException("Execution timed out");
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonEvaluationSupplier.java:58]: has nothing to do with time out, has something to do with script/expression execution failed.

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java [hunk 12]
```diff
@@ -218,34 +251,31 @@ private Serializable processReturnResult(EvaluationResults results) throws JsonP
...
+                throw new RuntimeException("Failed to run script.");
+            }
+            throw new RuntimeException("Timeout has been reached.");
```
Comment (lucian-cm) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:276]: this exception is missing the _t_ cause.

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java (hunk 5 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java]: done

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java (hunk 6 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java]: done

File: runtime-management/runtime-management-impl/src/test/java/io/cloudslang/runtime/impl/python/PythonExecutorTest.java (hunk 7 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/test/java/io/cloudslang/runtime/impl/python/PythonExecutorTest.java]: done

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java (hunk 12 above)
Comment (ilasalexandra) [reply runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonExecutor.java:277]: I renamed 't' to 'exception', and here 'exception' is not accessible. A RuntimeException will be thrown with the cause TimeoutException the same as in line 275

**Review by lucian-cm — APPROVED** [review]

**Pull request merged**